### PR TITLE
Delineation of responsibilities

### PR DESCRIPTION
**Pull request by akasprzok**
Introduces the notion of `ingest_api` and `client`.

`ingest_api` is responsible for formatting messages correctly for
different APIs such as humio-structured, humio-unstructured, hec, etc.,
as well as knowing how to construct headers and the path for those
specific APIs.

`client` is responsible for sending the messages formatted by
`ingest_api` to Humio.  Currently the only clients are a test client and
a Tesla client.  There are currently no plans to expand that.

### CODE REVIEW

**Comment by gburd** [comment]
LGTM

File: lib/client.ex [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  Defines the HTTP client interface used to send messages to the Humio ingest APIs
+  """
+  @callback send(
+              params :: %{
+                base_url: String.t(),
+                path: String.t(),
+                body: String.t(),
+                headers: list(tuple)
+              }
+            ) :: {:ok, response :: map()} | {:error, reason :: String.t()}
```
Comment (codeadict) [inline lib/client.ex:12]: Very little nitpick, shouldn't reason be `any()` here to match the Tesla specs?
```suggestion
            ) :: {:ok, response :: map()} | {:error, reason :: any())}
```

File: lib/ingest_api.ex [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  such as humio-structured, humio-unstructured, HEC, etc.
+  """
+  @callback transmit(
+              params :: %{
+                entries: nonempty_list(String.t()),
+                host: String.t(),
+                token: String.t(),
+                client: Logger.Backend.Humio.Client
+              }
+            ) :: {:ok, response :: map()}
```
Comment (codeadict) [inline lib/ingest_api.ex:13]: If this calls the client, it should also have the same return specs for errors 
```suggestion
            ) :: {:ok, response :: map()} | {:error, reason :: any())}
```

**Review by codeadict — APPROVED** [review]

**Review by larszen — COMMENTED** [review]
Looks good.

**Pull request merged**